### PR TITLE
fix(auth): log why has_permission denies instead of swallowing silently

### PR DESCRIPTION
**Pull request by kwit75**
## What
`TaskConn.has_permission` caught `PermissionError` from `resolve_team_permissions` and returned a bare `False`. Now it logs the reason first. **Behaviour is unchanged — it still denies.**

## Why
`resolve_team_permissions` raises when the session's `defaultTeam` isn't inside `account_info.organization` (AccountInfo carries a **single** org). The swallow meant the caller only saw `Permission 'x' denied`, with the actual cause — an unresolvable default team — invisible.

That is exactly what happened in saas **#373**: the admin Dashboard hung on *"Loading Dashboard Data"* with `Permission 'task.monitor' denied` for a user who **genuinely held** `task.monitor`. With no log line, it took ~2 days to trace.

## Related
- Root cause fix (backend): rocketride-saas **#374** — reconciles `default_team` to the session's selected org.
- Incident: rocketride-saas **#373**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved diagnostics for permission checks by logging a warning when default team permissions can’t be resolved.
  * Permission access continues to be denied safely when permission verification fails, with the warning detailing the user, team, requested permissions, and the denial reason.
* **Tests**
  * Added unit test coverage to confirm a single warning is emitted when permission resolution raises a `PermissionError`.

### CODE REVIEW

**Review by anandray — DISMISSED** [review]
**Verdict: approve. Surgical fix with one substantive concern about log level that should probably be resolved before merge.**

### What's good

- **Behavior preservation is explicit and correct.** Still `return False` in the exception path. The comment even says "Deny — but never silently."
- **Log message captures exactly the info that was missing during the incident**: `defaultTeam`, `userId`, denied `perm`, and the exception itself. That's everything an operator needs to jump straight to "reconcile default_team" (which is what saas #374 does).
- **`!r` for consistent representation** across log lines. `'foo'` stays quoted; `None` stays `None`. Grep-friendly.
- **Cross-refs the incident and the paired PR (#374)** in the comment. Future reader who lands on this line during a triage sees the whole story without spelunking.
- **`from rocketlib import debug` at module top** — consistent style, no local-scope import shenanigans.

### Substantive concern

**Log level: `debug`.**

The premise of this PR is "the swallow made #373 take two days to trace, so let's log the reason." If `rocketlib.debug` is filtered out of prod (which is what "debug" conventionally means in most log frameworks), then this fix has the same operational blast radius as the current code: silent in prod, only visible when someone happens to be running dev-level logging.

Two questions to resolve:

1. **What log level does `rocketlib.debug` map to in the deployed server config?** If it's always-on regardless of environment, my concern evaporates and the level is fine.

2. **If it IS environment-conditional, should this be `info` or `warning`?** Argument for `warning`: this branch fires when a user is being denied a permission the incident writeup says they *actually hold*. That's a real bug, just one that self-heals per saas #374 — anyone reading a warning-level log line will trace it faster than someone who doesn't see it at all. Argument for `info`: it's a notable-but-not-broken event once saas #374 lands.

I'd land on `warning` for the immediate future (while saas #374 is fresh and edge cases may still slip through) and downgrade to `info` after a stability window if the volume is high. `debug` reads like it defeats the purpose of the fix.

### Nit

- No test. Testing "did we emit a log line" is low-value; skip.

If the answer to Q1 above is "debug is always-on in prod for this service," I'm entirely comfortable — approving on that assumption. If not, please bump the level before merge.

**Comment by kwit75** [comment]
Checked it against the running cloud ALB rather than reasoning about it, and the answer kills the original approach twice over. Good question to have asked.

**1. `debug()` is not enabled in prod.** The engine command has no trace flag:

```
./engine ./ai/eaas.py --saas --host=0.0.0.0 --port=5565 --modelserver=wss://model.rocketride.dev:443
```

and `rocketlib.debug` is documented as "shown when the appropriate trace level is enabled". Corroborated rather than inferred: `debug(f"[account] upsert_user: ...")` fires on every single login, and appears **zero** times in the log.

**2. Bumping to `warning()` does not fix it** — which is what I would have shipped if I had just taken the suggestion. rocketlib output goes to the engine *job log*, which is delivered to the connected client, not to container stdout. A 5000-line sample of the prod ALB log is 100% uvicorn access lines:

```
1808  GET /            200
 681  GET /version     200
 237  connection open
 216  WebSocket /task/service [accepted]
```

Nothing from the engine at all, and no engine log file anywhere in the pod. So this PR as written delivered **no server-side observability whatsoever** — it would have looked like a fix and produced silence. And a denial reported only to the client being denied is not observability regardless of the level.

So: stdlib `logging`, which those uvicorn lines prove reaches stdout and therefore CloudWatch. There is existing precedent for it in `packages/ai` (`ai/node.py`, `common/models/vision/detection.py`) so this is not a new convention.

Volume is bounded by actual denials, and a flood of them is itself the signal — that is exactly the shape #373 had.

Pushed as 3be90c65.

Separately, this seems worth a follow-up beyond this PR: if engine-side `warning()`/`error()` never reach container stdout, then *no* server-side warning from any node is currently visible in prod. I have not confirmed how wide that is — happy to open an issue to scope it if you agree it looks that way.

**Review by anandray — DISMISSED** [review]
**Re-review approve.** Root-caused rather than paper-caused. Switching to stdlib `logging.getLogger(__name__).warning()` is the right answer — I would have shipped a level bump on rocketlib and it would still have been invisible. The inline comment naming both the trace-gate and the client-only sink means the next reader doesn't have to re-derive why it's stdlib.

**Comment by dmitriik-blip** [comment]
Added the regression assertion — `test_has_permission_logs_why_it_denied`, commit 93d7201a.

The point stands and I want to be clear I agree with it: the old test asserted only the `False` return, so it passed just as happily with the logging deleted. The log **is** the change, and nothing was holding it. The new test asserts the record carries the team, the user, the requested permission and the original `PermissionError` text, so dropping the line or dropping a field from it fails.

One correction on the mechanism. The suggestion was to mock `task_conn.debug` — that is stale as of this PR. I moved the call to stdlib `logging` precisely because `rocketlib`'s `debug()` never reaches container stdout in the cloud: it is trace-gated and prod runs without `--trace`, and rocketlib output goes to the engine job log delivered to the *connected client*, not to stdout. A 5000-line sample of the prod ALB log is 100% uvicorn access lines with nothing from the engine at all. Mocking `debug` would therefore assert against a call that no longer exists. The test uses `caplog` against the module logger instead.

**Two things I could not verify, stated plainly rather than left implied.**

I could not run it locally. The suite needs the built engine — pytest walks `fastapi` → `uvicorn` → `dotenv` → `rocketride`, at which point pip starts trying to fetch our own published package over the local source, and that is the wrong direction, so I stopped. I checked the assertions statically against the f-string in `task_conn.py` and all four substrings appear in it, but static agreement is not a passing test.

And CI will not run it either. `os-matrix.yml` is the only workflow invoking `./builder test`, and it fires on `workflow_dispatch` or a monthly cron — so `packages/ai/tests` gates nothing on a PR. A test added today first executes on the 1st of next month at the earliest. That is not os-matrix's fault, it is doing cross-OS compile verification and is meant to be expensive and rare; the gap is that there is no cheap per-PR job running just the Python unit tests on one platform. Filed as #1692 rather than widening this PR — it is the third instance today of tests existing with nothing running them, after rocketride-saas #400/#401.

So: the assertion is written and reviewable, and it is honestly unexecuted. If you would rather hold this PR until #1692 gives it somewhere to run, say so and I will.

File: packages/ai/src/ai/modules/task/task_conn.py [hunk 1]
```diff
@@ -381,7 +382,16 @@ def has_permission(self, perm: Union[list, str]) -> bool:
...
+        except PermissionError as exc:
+            # The session's defaultTeam is not resolvable inside account_info.organization
+            # (AccountInfo carries a single org). Deny — but never silently: without this
+            # line the caller sees a bare "Permission 'x' denied" and the real cause, an
+            # unresolvable default team, is invisible in the logs.
+            debug(
+                f'[auth] has_permission: cannot resolve defaultTeam '
+                f'{self._account_info.defaultTeam!r} for user {self._account_info.userId!r} '
+                f'-> denying {perm!r} ({exc})'
+            )
```
[the remark this answers is not in the document]
Comment (kwit75) [reply packages/ai/src/ai/modules/task/task_conn.py:407]: Done in 93d7201a — and the point was right: the old test asserted only the `False` return and would have passed with the logging deleted.

One correction on the mechanism, since it matters for anyone reading this later: the suggestion to mock `task_conn.debug` is stale as of this PR. The whole change is moving off `rocketlib.debug` onto stdlib `logging`, because `debug()` is trace-gated and prod runs without `--trace`, and rocketlib output goes to the engine job log delivered to the connected client rather than to container stdout. A 5000-line sample of the prod ALB log is 100% uvicorn lines with nothing from the engine in it. Mocking `debug` would assert against a call that no longer exists. The test uses `caplog` against the module logger instead.

Follow-up in 1e2a57a5 tightened it further — see the other thread.

File: packages/ai/tests/ai/modules/task/test_task_conn.py [hunk 2]
```diff
@@ -284,6 +286,32 @@ def _raise(info, team):
...
+    monkeypatch.setattr(tc_mod, 'resolve_team_permissions', _raise)
+
+    conn = _make_conn(account_info=_make_account_info(user_id='user-1', default_team='team-1'))
+    with caplog.at_level(logging.WARNING, logger=tc_mod.__name__):
+        assert conn.has_permission('task.control') is False
+
+    assert caplog.records, 'denial produced no log record at all'
+    message = ' '.join(r.getMessage() for r in caplog.records)
+    for expected in ('team-1', 'user-1', 'task.control', 'not in organization'):
+        assert expected in message, f'denial log omits {expected!r}: {message}'
```
[the remark this answers is not in the document]
Comment (kwit75) [reply packages/ai/tests/ai/modules/task/test_task_conn.py:320]: Fixed in 1e2a57a5, and this was a fair hit — my replacement test had the same weakness I was criticising in the original.

Joining every record meant an unrelated log could satisfy the field checks, and a level change would have gone unnoticed. That second part matters more than it looks here: this PR exists *because* the previous logging never reached stdout, so a silent downgrade to debug is precisely the regression that would make the message invisible again while the test stayed green.

Now filters to `tc_mod.__name__` at `logging.WARNING` and asserts exactly one record, so wrong logger, wrong level, missing, or duplicated all fail — and the assertion message prints what was actually captured so a failure is diagnosable.

**Review by anandray — APPROVED** [review]
**Re-review approve.** The new `test_has_permission_logs_why_it_denied` is exactly the shape my earlier review was groping at — pins the WARNING level (which determines whether CloudWatch sees it) and pins the field set (`team-1 / user-1 / task.control / not in organization`), so any silent-swallow regression trips the test. The docstring naming the tautology risk explicitly ("asserting only the False return is not enough: that assertion passes just as happily with the logging deleted, which is how the #373 incident stayed invisible for a day") is the right way to teach the next writer.

The filter `records = [r for r in caplog.records if r.name == tc_mod.__name__ and r.levelno == logging.WARNING]` catches the four regression shapes that would silently defeat the fix: downgrade to debug/info (0 WARNING records), upgrade to error (0 at this level), deletion (0), unrelated log added (2). Belt and braces.

Approving.
